File: pykde-bugs/xml_toolbar.cc
#include <qvbox.h>
#include <kapplication.h>
#include <kcmdlineargs.h>
#include <kmainwindow.h>
#include <ktoolbar.h>
#include <kxmlguiclient.h>
#include <kxmlguibuilder.h>
#include <kxmlguifactory.h>
#include <kaction.h>
#include <kactioncollection.h>

class MyVBox : public QVBox, public KXMLGUIClient
{
    Q_OBJECT;

    public:
      MyVBox(QWidget *parent);
      ~MyVBox();
      void createGUI();

    public slots:
      void noop();

    private:
      KToolBar *toolbar;
};

MyVBox::MyVBox(QWidget *parent) : QVBox(parent), KXMLGUIClient()
{
    toolbar = new KToolBar(this, "toolbar");

    KActionCollection *ac = actionCollection();
    new KAction("Action 1", 0, this, SLOT( noop() ), ac, "action1");
    new KAction("Action 2", 0, this, SLOT( noop() ), ac, "action2");
    new KAction("Action 3", 0, this, SLOT( noop() ), ac, "action3");

    setXMLFile("/tmp/xml_toolbar.rc");
    createGUI();
}

void
MyVBox::createGUI()
{
    KXMLGUIBuilder builder(this);
    KXMLGUIFactory factory(&builder, this);
    factory.addClient(this);
}

void
MyVBox::noop()
{
}

MyVBox::~MyVBox() 
{
    delete toolbar;
}

///////////////////////////////////////

class MainWindow : public KMainWindow
{
    Q_OBJECT;

    public:
      MainWindow();
      ~MainWindow();

    private:
        MyVBox *vbox;
};

MainWindow::MainWindow() : KMainWindow()
{
    vbox = new MyVBox(this);
}

MainWindow::~MainWindow()
{
    delete vbox;
}

///////////////////////////////////////

int
main (int argc, char **argv)
{
    KCmdLineArgs::init(argc, argv, "test", "test", "test", "1.0");
    KApplication app;
    MainWindow *window = new MainWindow;

    app.setTopWidget(window);
    window->show();
    return app.exec();
}

#include "xml_toolbar_moc.cc"

File: pykde-bugs/xml_toolbar.py
#! /usr/bin/env python

import sys

import qt
import kdeui
import kdecore

def main():
    application = kdecore.KApplication(sys.argv, 'test app')
    main_window = MainWindow(None, 'main window')
    main_window.show()
    application.exec_loop()

class MainWindow(kdeui.KMainWindow):
    def __init__ (self, *args):
        kdeui.KMainWindow.__init__(self, *args)
        self.vbox = MyVBox(self)

class MyVBox(qt.QVBox, kdeui.KXMLGUIClient):
    def __init__(self, parent):
        qt.QVBox.__init__(self, parent)
        kdeui.KXMLGUIClient.__init__(self)
        self.toolbar = kdeui.KToolBar(self, 'toolbar')

        # This assert succeeds...
        assert 'actionCollection' in dir(self)

        # ... but this does not:
        print hasattr(self, 'actionCollection')

        # XXX ... and this obviously dies with AttributeError
        # Using super() instead of calling each constructor does not help
        ac = self.actionCollection()

        kdeui.KAction('Action 1', kdecore.KShortcut.null(), self.noop, ac, 'action1')
        kdeui.KAction('Action 2', kdecore.KShortcut.null(), self.noop, ac, 'action2')
        kdeui.KAction('Action 3', kdecore.KShortcut.null(), self.noop, ac, 'action3')

        self.setXMLFile('/tmp/xml_toolbar.rc');
        self.createGUI()

    def createGUI(self):
        builder = kdeui.KXMLGUIBuilder(self)
        factory = kdeui.KXMLGUIFactory(builder, self)
        factory.addClient(self)

    def noop(self):
        pass

if __name__ == '__main__':
    main()

